File: src/main/java/com/cuupa/mailprocessor/services/archive/FileProtocol.java
package com.cuupa.mailprocessor.services.archive;

import java.util.List;

public interface FileProtocol extends AutoCloseable {

    void init(String username, String password);

    boolean exists(String path);

    boolean save(String path, byte[] data);

    boolean createDirectory(String path);

    List<ArchiveResource> list(String path);
}

File: src/main/java/com/cuupa/mailprocessor/services/archive/FileProtocol.kt
package com.cuupa.mailprocessor.services.archive

import java.io.InputStream

interface FileProtocol : AutoCloseable {

    fun init(username: String?, password: String?)

    fun exists(path: String, filename: String): Boolean

    fun save(path: String, filename: String, data: ByteArray): Boolean

    fun createDirectory(path: String): Boolean

    fun list(path: String): List<FileResource>

    fun get(path: String, filename: String): InputStream

    fun delete(path: String, filename: String): Boolean

    fun createDirectories(url: String, path: String): String
}